diceware.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "DICEWARE" "1" "January 2026" "diceware 1.1.0.dev0" "User Commands"
.SH NAME
diceware \- create passphrases
.SH SYNOPSIS
.sp
\fBdiceware\fP [\fIOPTION\fP]... [\fIFILE\fP]
.SH DESCRIPTION
.sp
\fBdiceware\fP generates passphrases by concatenating words randomly picked from
wordlists. It supports also real dice for passphrase generation.
.sp
It is based on the proposals of Arnold G. Reinhold on  <http://diceware.com> \&.
.SH OPTIONS
.sp
\fBpositional arguments\fP:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.B FILE
optional input wordlist. \fB\(aq\-\(aq\fP will read from stdin. Should contain one
word per line.
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fBoptional arguments\fP:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.B \fB\-h\fP, \fB\-\-help\fP
show help message and exit
.TP
.B \fB\-n\fP \fINUM\fP, \fB\-\-num\fP \fINUM\fP
number of words to concatenate. Default 6
.TP
.B \fB\-c\fP, \fB\-\-caps\fP
Capitalize words. This is the default.
.TP
.B \fB\-\-no\-caps\fP
Turn off capitalization.
.TP
.B \fB\-s\fP \fINUM\fP, \fB\-\-specials\fP \fINUM\fP
Append NUM special chars at the end of the generated word.
.TP
.B \fB\-d\fP \fIDELIMITER\fP, \fB\-\-delimiter\fP \fIDELIMITER\fP
Separate words by DELIMITER. Empty string by default.
.TP
.B \fB\-r\fP \fISOURCE\fP, \fB\-\-randomsource\fP \fISOURCE\fP
Get randomness from this source. Possible values:
\fBrealdice\fP, \fBsystem\fP\&. Default: \fBsystem\fP
.TP
.B \fB\-w\fP [\fINAME\fP [\fINAME\fP ...]], \fB\-\-wordlist\fP [\fINAME\fP [\fINAME\fP ...]]
Use words from this wordlist. Possible values: \fIca\fP, \fIde\fP, \fIde_8k\fP, \fIen\fP,
\fIen_8k\fP, \fIen_adjectives\fP, \fIen_eff\fP, \fIen_nouns\fP, \fIen_orig\fP, \fIen_securedrop\fP\&.
\fIes\fP, \fIit\fP, \fIpt\-br\fP\&. Default: \fBen_eff\fP
.TP
.B \fB\-v\fP, \fB\-\-verbose\fP
Be verbose. Use several times for increased verbosity.
.TP
.B \fB\-\-version\fP
output version information and exit.
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fBArguments related to\fP \fIrealdice\fP \fBrandomsource\fP:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.B \fB\-\-dice\-sides\fP \fIN\fP
Number of sides of dice. Default: 6
.UNINDENT
.UNINDENT
.UNINDENT
.SH ENVIRONMENT VARIABLES
.INDENT 0.0
.TP
.B \fBXDG_CONFIG_HOME\fP
If set and not empty, this variable determines the directory to use for
user\-local configuration files. We then lookup
\fI${XDG_CONFIG_HOME}/diceware/diceware.ini\fP and values set here override
system\-wide config files.
.TP
.B \fBXDG_CONFIG_DIRS\fP
If set and not empty, this variable is interpreted as colon\-separated list
of directories, that might contain system\-wide configuration files. We
lookup \fI<DIR>/diceware/diceware.ini\fP for each directory set in
\fI$XDG_CONFIG_DIRS\fP\&.
.TP
.B \fBXDG_DATA_HOME\fP
.INDENT 7.0
.INDENT 3.5
If set and not empty, this variable determines a directory to search for
additional wordlists. We then lookup \fI${XDG_DATA_HOME}/diceware\fP for any
existing wordlist files.
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B \fBXDG_DATA_DIRS\fP
If set and not empty, this variable is interpreted as colon\-separated list
of directories, that might contain additional wordlist files. See below. We
lookup \fI<DIR>/diceware/\fP then for each directory set in the list.
.UNINDENT
.UNINDENT
.SH FILES
.sp
Depending on environment variables set (or not set) we lookup certain
directories for configuration files called \fBdiceware.ini\fP and for wordlist
files.
.SS CONFIGURATION FILES
.sp
Configuration settings for \fBdiceware\fP can be spread over several
configuration files. We parse configuration values from the files given below,
but values set in former files take precedence over values set in latter ones.
.INDENT 0.0
.TP
.B \fI~/.diceware.ini\fP
Your personal diceware configuration file. Values set here override values
from any other configuration file.
.TP
.B \fI$XDG_CONFIG_HOME/diceware/diceware.ini\fP
Additional location for your personal diceware configuration. Values set
here will override any system\-wide valid values but can be overridden by
\fI~/.diceware.ini\fP\&.
.TP
.B \fI$HOME/.config/diceware/diceware.ini\fP
Alternative location for diceware configuration, only used if
\fI${XDG_CONFIG_HOME}\fP is empty or unset.
.TP
.B \fI/etc/xdg/diceware/diceware.ini\fP
If \fB$XDG_CONFIG_DIRS\fP is not set or empty, we look here for a system\-wide
configuration file. Values set here take least precedence.
.UNINDENT
.SS WORDLIST FILES AND WORDLIST DIRECTORIES:
.sp
\fBdiceware\fP comes with a set of wordlists but enables you to add new wordlists
by putting them into certain directories. The paths where the lists are stored
(including the built\-in ones) is shown using \fB\-\-show\-wordlist\-dirs\fP\&.
.sp
Wordlist files are expected to contain lines with one term on each
line and they must have a certain filenames to be found.
.sp
Wordlist filenames have to follow the pattern: \fBwordlist_<NAME>.txt\fP
where \fB<NAME>\fP can be any name consisting of letters, numbers, underscores and
hyphens. For instance \fBwordlist_en_eff.txt\fP is the filename of the EFF
(electronic frontier foundation) word list. \fBen_eff\fP is the name of this list.
.sp
We support \fB\&.txt\fP and \fB\&.asc\fP as filename extensions for wordlists, where
\fB\&.txt\fP files are expected to be plain wordlists and \fB\&.asc\fP files should
provide a PGP\-signature.
.sp
If wordlists with the same name are found in different directories then the one
in the directory with the highest precedence is taken only. The following
locations are ordered by precedence (highest first). Therefore built\-in
wordlists cannot be overridden by custom wordlists. You can, however, use
custom wordlists with a different name.
.sp
Directories we look up that do not exist (in part or completely) are silently
skipped when searching for wordlist files.
.INDENT 0.0
.TP
.B \fI<INSTALL\-DIR>/wordlists/\fP
The directory containing the built\-in wordlists as part of the
installation. These are the wordlists that are always available, regardless
of configuration values and their exact location depends on the
installation location of the \fBdiceware\fP package.
.TP
.B \fI$XDG_DATA_HOME/diceware/\fP
If $XDG_DATA_HOME is set and not empty, we look in this directory for
wordlists.
.TP
.B \fI$HOME/.local/share/diceware/\fP
If $XDG_DATA_HOME is unset or empty, we look  into this directory for
wordlists.
.TP
.B \fI<DIR>/diceware\fP from \fI$XDG_DATA_DIRS\fP
If $XDG_DATA_DIR is set and not empty, it is interpreted as a
colon\-separated list of directories with \fI/diceware\fP appended. So,
\fI/foo/bar:/baz\fP will make us look into \fI/foo/bar/diceware/\fP and
\fI/baz/diceware/\fP in that order.
.TP
.B \fI/usr/local/share/diceware/\fP, \fI/usr/share/diceware\fP
If $XDG_DATA_DIRS is unset or empty, we look into these two directories for
wordlists.
.UNINDENT
.SH EXAMPLES
.INDENT 0.0
.TP
.B \fBdiceware\fP
Create a passphrase using defaults. Outputs something like
\(dq\fBWheelDyeHonkCanvasWitsPuck\fP\(dq
.TP
.B \fBdiceware \-d\fP \fI\(dq\-\(dq\fP \fB\-n\fP \fI3\fP
Create a passphrase with three words, separated by dash (\(dq\fI\-\fP\(dq). Results in
something like \(dq\fBWheel\-Dye\-Honk\fP\(dq
.TP
.B \fBdiceware \-\-no\-caps\fP
Create a passphrase without capital words. Creates something like
\(dq\fBwheel\-dye\-honk\fP\(dq.
.TP
.B \fBdiceware \-r\fP \fIrealdice\fP
Use real dice to create a passphrase. The program will tell you what to do
(roll dice and tell what numbers appear) and in the end present a
passphrase.
.TP
.B \fBdiceware \-r\fP \fIrealdice\fP \fB\-\-dice\-sides\fP \fI20\fP
Use real dice, as shown above, but this time use dice with 20 faces,
instead of standard, 6\-sided dice.
.TP
.B \fBdiceware mywordlist.txt\fP
Create a passphrase with words from file \(dqmywordlist.txt\(dq. The file should
contain one word on each line.
.TP
.B \fBdiceware \-w en_securedrop \-s 2\fP
Create a passphrase with two special chars appended at the end of the
generated passphrase and containing words from wordlist
\(dq\fBen_securedrop\fP\(dq. This is one of the wordlists that come included with
\fIdiceware\fP\&. Creates something like:
\(dq\fBPlayaBrigVertSeesNestsGets{\-\fP\(dq.
.TP
.B \fBdiceware \-w en_adjectives en_nouns \-n 2\fP
Create two syntactically meaningful phrases, each one consisting of an
adjective and a noun. Results in something like:
\(dq\fBCruelAttendeesCleanCoffee\fP\(dq.
.UNINDENT
.SH COPYRIGHT
.sp
Copyright (C) 2015\-2026 Uli Fouquet and contributors
.sp
This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.
.sp
This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.
.sp
You should have received a copy of the GNU General Public License along with
this program.  If not, see < <http://www.gnu.org/licenses/> >.
.sp
diceware is a concept invented by Arnold G. Reinhold, Cambridge, Massachusetts
USA.
.sp
The Securedrop wordlist (file \fBwordlists/wordlist_en_securedrop.asc\fP) by
Heartsucker is licensed under the \fIMIT\fP license (see  <http://mit\-license.org/> ).
.sp
The EFF wordlist (file \fBwordlsts/wordlist_en_eff.txt\fP) is licensed by the
Electronic Frontier Foundation under the \fICreative Commons CC\-BY 3.0 US\fP
license (see  <https://creativecommons.org/licenses/by/3.0/us/> ).
.sp
The copyright for the the \fIDiceware SecureDrop\fP list is owned by \fI@heartsucker\fP\&.
Copyright for the \fIEFF large\fP list by \fIJoseph Bonneau\fP and \fIEFF\fP\&. Copyright for
the brazilian portuguese list by \fI@drebs\fP\&. Copyright for the english adjective
and noun lists by \fINaturalLanguagePasswords\fP\&.
.sp
\(dqDiceware\(dq is a trademark of Arnold G Reinhold, used with permission.
.SH AUTHOR
Written by Uli Fouquet and contributors
.\" Generated by docutils manpage writer.
.


CHANGES.rst
Changes
=======

1.1.0.dev0 (yet unreleased)
---------------------------

- Set prog name for `ArgumentParser` explicitly to work around difficult to
  predict programme-name computation changed in Python3.14. Fixed #122.
- Change special-char-behaviour: instead of pouring special chars over the
  generated passphrase and thus rendering it glibberish and unreadable (and
  harder to remember), we now append them always at the end of the
  passphrase. Fixed #116.
- Update dependency list for generating docs.


1.0.1 (2024-12-24)
------------------

Security Upgrade.

- Update dependency list for generating docs. The old dependencies required
  an unsafe version of ``jinja2``.


1.0 (2024-12-24)
----------------

Major overhaul of the whole project. We introduce more modern approaches in
project layout (like ``pyproject``), use new linters and other tools while we
still support all Python versions from 2.7 up to current 3.12.

We now follow the XDG base directory specification, that tells where config
(and other) files for an application can be found. You now can also use
``${XDG_CONFIG_HOME}/diceware/diceware.ini``, or, if the given var is emtpy or
unset, ``${HOME}/.config/diceware/diceware.ini``. The traditional location
``${HOME}/.diceware.ini`` is still supported.

Furthermore we read ``${XDG_DATA_HOME}/diceware/`` (or
``${HOME}/.local/share/diceware/`` if ``${XDG_DATA_HOME}`` is empty or unset) to
lookup further wordlists.

New option ``--show-wordlist-dirs`` lists all directory locations we search for
contained wordlists.

- Officially support Python 3.10 to 3.12.
- Fixed #86: Follow `XDG <https://specifications.freedesktop.org/basedir-spec/latest/>`_ base directory specification.
- Fixed #84: Allow to store wordlists also in custom directories.
- Use ``ruff`` as linter, drop ``flake8``.
- Renew ``tox`` configuration.
- Switch to ``pyproject``-based project layout, away from using ``setup.py``.
- Fixed #62: Removed ``pkg_resources`` dependency. Kudos to @htgoebel!
  As a side effect no randomness sources from other packages are supported anymore.
- Added French wordlist. Many thanks for the good work go to Tango.
- Added Catalan, Spanish and Italian wordlists.


0.10 (2022-02-15)
-----------------

- Officially support Python 3.8 and Python 3.9.
- Removed official support for pypy2, Python 2.6, and Python 3.3.
- Allow to specify several wordlists in order to create syntactical valid
  phrases.
- Also added first wordlists with english adjectives/nouns to generate for
  instance <adjective-noun> phrases that are easier to memorize.
- When using real dice, allow entering of several rolls at once. Patch from
  `Adin Hoyle <alan@alanhoyle.com>`_.
- Added german wordlists.
- Added carefully compiled brazilian portugese wordlist. Kudos to @drebs.
- Removed original diceware wordlists temporarily, for containing bad terms (#85)


0.9.6 (2018-12-19)
------------------

- Officially support Python 3.7.
- Fixed #51: Fix to formatting of list in Wordlists section. Kudos to `Doug
  Muth <https://github.com/dmuth>`_.


0.9.5 (2018-04-07)
------------------

- Fixed #28: use Debian-compliant, machine-readable copyright format. Kudos to
  `Simon Fondrie-Teitler <https://github.com/simonft>`_ and @anarcat.
- Fixed #48: Clarify trademark status of diceware. Mr. Reinhold granted
  permission to use the name 'Diceware' in this project and under the conditions
  listed in the issue comments. Many thanks to him!


0.9.4 (2018-02-27)
------------------

- Set default logging level to ERROR (was: CRITICAL)
- Fixed #44: provide a short and readable file-not-found message (many thanks to
  `bhavin192 <https://github.com/bhavin192>`_)
- Fixed #45: clean up logging handlers after test runs.
- Removed date-dependent tests from default test suite. Run ``py.test -m ''``
  or ``tox`` to run them.


0.9.3 (2017-09-14)
------------------

- Fix broken test.


0.9.2 (2017-09-14)
------------------

- Fixed #33. Make ``en_eff`` the new default wordlist. This results in slightly
  decreased entropy per word (12.92 bits instead of 13.0), but provides prefix
  code and better memorizable words. Thanks to @anarcat for the suggestion.
- Fixed #35. Make ``realdice`` source of randomness provide an equal distribution
  of roll numbers even for sequences shorter than number of dice sides.
- Added a man page.
- Support Python 3.6.
- Import ``ConfigParser`` instead of ``SafeConfigParser`` if the latter is an alias
  of the former.
- Fixed #37. Ensure file descriptors are closed properly.
- Fixed #38. Get wordlists dir by function (instead of const) to allow
  reproducible builds. Kudos go to @drebs, again.


0.9.1 (2016-12-24)
------------------

- Fixed #32, in docs tell that ``--no-caps`` option does not generate
  lower-case terms.
- Fixed #31, broken ``realdice`` source of randomness. ``argparse`` related bug,
  Bug was discovered and fixed by @LogosOfJ, thanks a lot!
- Fixed #29. Tell about code prefix problem in README.
- Activated logging. Using ``verbose`` will result in additional output.


0.9 (2016-09-14)
----------------

- Added ``--dice-sides`` option to tell how many sides used dices
  provide.
- Changed API interface of ``get_config_dict()`` to allow more flexible
  handling of config files.
- Support different verbosity levels.
- Added new wordlist ``en_eff``. It is a 7776-terms list provided by
  the Electronic Frontier Foundation. See
  https://www.eff.org/deeplinks/2016/07/new-wordlists-random-passphrases
  for details. Thanks to `George V. Reilly
  <https://github.com/georgevreilly>`_ for hinting!
- Fixed #27. Allow dashes in numbered wordlists. Yet, these looked
  like ``1234 myterm``. We now also accept ``1-2-3-4 myterm``.


0.8 (2016-05-07)
----------------

- Closed #23. @dwcoder provided a fix that allows use of
  whitespace-only values in diceware config files if they are enclosed
  in quotes.
- Fixed #21. @dwcoder revealed and fixed (again!). This time ``--caps``

diceware.py
#  diceware -- passphrases to remember
#  Copyright (C) 2015  Uli Fouquet
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
import argparse
import os
import pkg_resources
import re
import sys
from random import SystemRandom

__version__ = pkg_resources.get_distribution('diceware').version

#: The directory in which wordlists are stored
SRC_DIR = os.path.dirname(__file__)

#: A regular expression matching 2 consecutive ASCII chars. We
#: consider this to represent some language/country code.
RE_LANG_CODE = re.compile('^[a-zA-Z]{2}$')

#: Special chars inserted on demand
SPECIAL_CHARS = "~!#$%^&*()-=+[]\{}:;\"'<>?/0123456789"


GPL_TEXT = (
    """
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
    """
    )


def print_version():
    """Output current version and other infos.
    """
    print("diceware %s" % __version__)
    print("Copyright (C) 2015 Uli Fouquet")
    print("diceware is based on suggestions of Arnold G. Reinhold.")
    print("See http://diceware.com for details.")

    print("'Diceware' is a trademark of Arnold G. Reinhold.")
    print(GPL_TEXT)


def handle_options(args):
    """Handle commandline options.
    """
    parser = argparse.ArgumentParser(description="Create a passphrase")
    parser.add_argument(
        '-n', '--num', default=6, type=int,
        help='number of words to concatenate. Default: 6')
    cap_group = parser.add_mutually_exclusive_group()
    cap_group.add_argument(
        '-c', '--caps', action='store_true',
        help='Capitalize words. This is the default.')
    cap_group.add_argument(
        '--no-caps', action='store_false', dest='capitalize',
        help='Turn off capitalization.')
    parser.add_argument(
        '-s', '--specials', default=0, type=int, metavar='NUM',
        help="Insert NUM special chars into generated word.")
    parser.add_argument(
        'infile', nargs='?', metavar='INFILE', default=None,
        type=argparse.FileType('r'),
        help="Input wordlist. `-' will read from stdin.",
        )
    parser.add_argument(
        '--version', action='store_true',
        help='output version information and exit.',
        )
    parser.set_defaults(capitalize=True)
    args = parser.parse_args(args)
    return args


def get_wordlist(file_descriptor):
    """Parse file in `file_descriptor` and build a word list of it.

    `file_descriptor` is expected to be a file descriptor, already
    opened for reading. The descriptor will be closed after
    processing.

    A wordlist is expected to contain lines of words. Each line a
    word. Empty lines are ignored. Returns a list of terms (lines)
    found.
    """
    result = [
        line.strip() for line in file_descriptor.readlines()
        if line.strip() != '']
    file_descriptor.close()
    return result


def get_wordlist_path(lang):
    """Get path to a wordlist file for language `lang`.

    The `lang` string is a 2-char country code. Invalid codes raise a
    ValueError.
    """
    if not RE_LANG_CODE.match(lang):
        raise ValueError("Not a valid language code: %s" % lang)
    basename = 'wordlist_%s.txt' % lang
    return os.path.abspath(os.path.join(
        SRC_DIR, 'diceware-wordlists', basename.lower()))


def insert_special_char(word, specials=SPECIAL_CHARS, rnd=None):
    """Insert a char out of `specials` into `word`.

    `rnd`, if passed in, will be used as a (pseudo) random number
    generator. We use `.choice()` only.

    Returns the modified word.
    """
    if rnd is None:
        rnd = SystemRandom()
    char_list = list(word)
    char_list[rnd.choice(range(len(char_list)))] = rnd.choice(specials)
    return ''.join(char_list)


def get_passphrase(wordnum=6, specialsnum=1, delimiter='', lang='en',
                   capitalized=True, fd=None):
    """Get a diceware passphrase.

    The passphrase returned will contain `wordnum` words deliimted by
    `delimiter`.

    If `capitalized` is ``True``, all words will be capitalized.

    If `fd`, a file descriptor, is given, it will be used instead of a
    'built-in' wordlist (and `lang` will be ignored).

    The wordlist to pick words from is determined by `lang`,
    representing a language.
    """
    if fd is None:
        fd = open(get_wordlist_path(lang), 'r')
    word_list = get_wordlist(fd)
    rnd = SystemRandom()
    words = [rnd.choice(word_list) for x in range(wordnum)]
    if capitalized:
        words = [x.capitalize() for x in words]
    result = delimiter.join(words)
    for x in range(specialsnum):
        result = insert_special_char(result, rnd=rnd)
    return result


def main(args=None):
    """Main programme.

    Called when `diceware` script is called.

    `args` is a list of command line arguments to process. If no such
    args are given, we use `sys.argv`.
    """
    if args is None:
        args = sys.argv[1:]
    options = handle_options(args)
    if options.version:
        print_version()
        raise SystemExit(0)
    print(get_passphrase(
        wordnum=options.num,
        specialsnum=options.specials,
        capitalized=options.capitalize,
        fd=options.infile
        )
    )
